src/loader/tsbgw_launcher.c

```diff
@@ -0,0 +1,511 @@
+	worker.bgw_main_arg = db_id;
+	memcpy(worker.bgw_extra, &vxid, sizeof(VirtualTransactionId));
+
+	return RegisterDynamicBackgroundWorker(&worker, handle);
+
+}
+
+static volatile sig_atomic_t got_sigterm = false;
+static void
+tsbgw_sigterm(SIGNAL_ARGS)
+{
+	int			save_errno = errno;
+
+	got_sigterm = true;
+	SetLatch(MyLatch);
+
+	errno = save_errno;
+}
+
+
+
+/*
+ * Model this on autovacuum.c -> get_database_list
+ * Note that we are not doing all the things around memory context that they do, because
+ * a) we're using shared memory to store the list of dbs and b) we're in a function and
+ * shorter lived context here.
+ * This can get called at two different times 1) when the cluster launcher starts and is looking for dbs
+ * and 2) if it restarts due to a postmaster signal.
+ */
+static HTAB *
+populate_database_htab(void)
+{
+	Relation	rel;
+	HeapScanDesc scan;
+	HeapTuple	tup;
+	HTAB	   *db_htab = NULL;
+	HASHCTL		info;
+
+	/*
+	 * first initialize the hashtable before we start a txn, we'll be writing
+	 * info about dbs there
+	 */
+	info = (HASHCTL)
+	{
+		.keysize = sizeof(Oid),
+			.entrysize = sizeof(tsbgwHashEntry)
+	};
+	db_htab = hash_create("tsbgw_launcher_db_htab", 8, &info, HASH_BLOBS | HASH_ELEM);
+
+
+	/*
+	 * by this time we should already be connected to the db, and only have
+	 * access to shared catalogs
+	 */
+	/* start a txn, see note in autovacuum.c for why */
+	StartTransactionCommand();
+	(void) GetTransactionSnapshot();
+
+	rel = heap_open(DatabaseRelationId, AccessShareLock);
+	scan = heap_beginscan_catalog(rel, 0, NULL);
+
+	while (HeapTupleIsValid(tup = heap_getnext(scan, ForwardScanDirection)))
+	{
+		Form_pg_database pgdb = (Form_pg_database) GETSTRUCT(tup);
+		tsbgwHashEntry *tsbgw_he;
+		Oid			db_oid;
+		bool		hash_found;
+
+		if (!pgdb->datallowconn)
+			continue;			/* don't bother with dbs that don't allow
+								 * connections, we'll fail when starting a
+								 * worker anyway */
+
+		db_oid = HeapTupleGetOid(tup);
+		tsbgw_he = (tsbgwHashEntry *) hash_search(db_htab, &db_oid, HASH_ENTER, &hash_found);
+		if (!hash_found)
+			tsbgw_he->db_scheduler_handle = NULL;
+
+
+	}
+	heap_endscan(scan);
+	heap_close(rel, AccessShareLock);
+	CommitTransactionCommand();
+	return db_htab;
+}
+
+static void
+start_db_schedulers(HTAB *db_htab)
+{
+	HASH_SEQ_STATUS hash_seq;
+	tsbgwHashEntry *current_entry;
+
+	/* now scan our hash table of dbs and register a worker for each */
+	hash_seq_init(&hash_seq, db_htab);
+
+
+	while ((current_entry = hash_seq_search(&hash_seq)) != NULL)
+	{
+		bool		worker_registered = false;
+		pid_t		worker_pid;
+		VirtualTransactionId vxid;
+
+		/* When called at server start, no need to wait on a vxid */
+		SetInvalidVirtualTransactionId(vxid);
+
+		worker_registered = register_tsbgw_entrypoint_for_db(current_entry->db_oid, vxid, &current_entry->db_scheduler_handle);
+		if (worker_registered)
+		{
+			WaitForBackgroundWorkerStartup(current_entry->db_scheduler_handle, &worker_pid);
```

Comment:
Yeah. But it's not a huge cost and I'd really like the handle being there to mean that it has started up. If that takes a long time, it's most likely going to mean that it wouldn't help to parallelize because we're just running out of available background workers.